pandas/tseries/frequencies.py

```diff
@@ -172,15 +176,12 @@ def infer_freq(
         inferer = _TimedeltaFrequencyInferer(index)
         return inferer.get_freq()
 
-    if isinstance(index, Index) and not isinstance(index, DatetimeIndex):
+    elif isinstance(index, Index):
         if is_numeric_dtype(index.dtype):
             raise TypeError(
                 f"cannot infer freq from a non-convertible index of dtype {index.dtype}"
             )
-        # error: Incompatible types in assignment (expression has type
-        # "Union[ExtensionArray, ndarray[Any, Any]]", variable has type
-        # "Union[DatetimeIndex, TimedeltaIndex, Series, DatetimeLikeArrayMixin]")
-        index = index._values  # type: ignore[assignment]
+        index = cast("TimedeltaArray", index._values)
```

Comment:
this cant be right.  TimedeltaIndex goes through the _TimedeltaFrequencyInferer path above

Reply:
Thank you for your comment. If I understand correctly, we only need to check` is_numeric_dtype(index.dtype) `then.

Reply:
yep, looks like that should work.

Reply:
thank you for reviewing this pr.